### ISSUE TITLE
Handle "Invalid Credentials" without throwing exception

### ISSUE THREAD

**Issue by tylerherwig** [Adldap2/Adldap2#692]
- Adldap2 Version:10.0.11
- LDAP Type: Active Directory
- PHP Version: 7.0.33

### Description:

I am attempting to write a small AD Auth based login for an application via an AJAX call. I am simply taking a login form and gathering the user's username & password and sending a POST request to the back-end where I am attempting to authenticate based on the credentials.

I have referenced the following example: https://adldap2.github.io/Adldap2/#/setup?id=authenticating
```

try {

	if ($provider->auth()->attempt($loginData['username'], $loginData['password'])) {
		echo json_encode(
			array (
				'login' => true
			)
		);
	} else {
		echo json_encode(
			array (
				'login' => false,
				'message' => 'Error message here.'
			)
		);
	}
} catch (Adldap\Auth\UsernameRequiredException $e) {
	// The user didn't supply a username.
} catch (Adldap\Auth\PasswordRequiredException $e) {
	// The user didn't supply a password.
}
```

If successful, I am returning a true value to my frontend. This is working as intended. However, on the case of someone entering a wrong password or username...

It reaches the else statement and it's throwing an exception. Is there a way to avoid throwing an exception and instead handle in my own manor based on the error thrown?

**Comment by stevebauman** [comment]
Hi @tylerherwig,

> It reaches the else statement and it's throwing an exception.

What exception is it throwing?

**Comment by tylerherwig** [comment]
It is throwing "Invalid Credentials" which is correct. The credentials are indeed invalid but I'd ideally like to handle this in the else statement differently rather than throwing an exception.

[27-Jun-2019 11:57:05 America/New_York] PHP Exception:  Invalid credentials in C:\inetpub\common_includes\AD\adldap2\vendor\adldap2\adldap2\src\Auth\Guard.php on line 104
[27-Jun-2019 11:57:05 America/New_York] PHP Stack trace:
[27-Jun-2019 11:57:05 America/New_York] PHP   1. {main}() C:\inetpub\wwwroot\secured\apps\production\dtc_audit_application\components\process.php:0
[27-Jun-2019 11:57:05 America/New_York] PHP   2. Adldap\Auth\Guard->attempt($username = 'NETWORK\\therwig', $password = 'password', $bindAsUser = *uninitialized*) C:\inetpub\wwwroot\secured\ apps\production\dtc_audit_application\components\process.php:15
[27-Jun-2019 11:57:05 America/New_York] PHP   3. Adldap\Auth\Guard->bind($username = 'NETWORK\\therwig', $password = 'password') C:\inetpub\common_includes\AD\adldap2\vendor\adldap2\adldap2\src\Auth\Guard.php:67
[27-Jun-2019 11:57:05 America/New_York] PHP Adldap\Auth\BindException:  Invalid credentials in C:\inetpub\common_includes\AD\adldap2\vendor\adldap2\adldap2\src\Auth\Guard.php on line 109
[27-Jun-2019 11:57:05 America/New_York] PHP Stack trace:
[27-Jun-2019 11:57:05 America/New_York] PHP   1. {main}() C:\inetpub\wwwroot\secured\apps\production\dtc_audit_application\components\process.php:0
[27-Jun-2019 11:57:05 America/New_York] PHP   2. Adldap\Auth\Guard->attempt($username = 'NETWORK\\therwig', $password = 'password', $bindAsUser = *uninitialized*) C:\inetpub\wwwroot\secured\ apps\production\dtc_audit_application\components\process.php:15
[27-Jun-2019 11:57:05 America/New_York] PHP   3. Adldap\Auth\Guard->bind($username = 'NETWORK\\therwig', $password = 'password') C:\inetpub\common_includes\AD\adldap2\vendor\adldap2\adldap2\src\Auth\Guard.php:67

**Comment by stevebauman** [comment]
That's strange, using the `auth()->attempt()` method, exception are caught and should not be generated during this call.

Warnings are also muted - as shown [here](https://github.com/Adldap2/Adldap2/blob/v10.0.11/src/Auth/Guard.php#L101).

Are you setting any type of error reporting via `error_reporting(E_ALL)` in any of your code?

**Comment by tylerherwig** [comment]
Yes, we are setting `error_reporting(E_ALL)` in our php.ini file.

We are using the recommended php.ini.

What do you recommend we change to be able to gracefully handle the exceptions? We've found that temporarily disabling error logging during this code logic and then re-enabling works, but this does not seem ideal...

`ini_set('log_errors', false);`
`ini_set('display_errors', 'Off');`

and...

`ini_set('log_errors', true);`
`ini_set('display_errors', 'On');`

Would it make more sense to return the exception rather than throwing the exception in code?
